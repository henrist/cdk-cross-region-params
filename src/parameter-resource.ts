import * as ssm from "@aws-cdk/aws-ssm"
import * as cdk from "@aws-cdk/core"
import { CrossRegionParameter } from "./cross-region-parameter"

type ReferenceToResource<T> = (
  scope: cdk.Construct,
  id: string,
  reference: string,
) => T

export interface ParameterResourceProps<T> {
  /**
   * The nonce can be used to force the stack to update to check
   * for a new value.
   *
   * @default Date.now().toString()
   */
  nonce?: string
  parameterName: string
  resource: T
  resourceToReference(resource: T): string
  referenceToResource: ReferenceToResource<T>
  /**
   * List of regions that can retrieve this resource from an SSM Parameter.
   */
  regions: string[]
}

/**
 * Register SSM Parameters in other regions storing a reference to
 * a resource, which can then be resolved in the other region by
 * reading from the parameter.
 *
 * Storing the SSM Parameters in the other regions speeds up the
 * resolving of parameters, since we can use CloudFormation SSM
 * Parameters instead of custom resources.
 *
 * If the resource is in the same region, the resource will be returned
 * like normally in CDK, causing an export/import if cross-stack.
 */
export class ParameterResource<T> extends cdk.Construct {
  private readonly nonce: string
  private readonly parameterName: string
  private readonly resource: T
  private readonly referenceToResource: ReferenceToResource<T>
  private readonly regions: string[]

  constructor(
    scope: cdk.Construct,
    id: string,
    props: ParameterResourceProps<T>,
  ) {
    super(scope, id)
    this.nonce = props.nonce ?? Date.now().toString()
    this.parameterName = props.parameterName
    this.resource = props.resource
    this.referenceToResource = props.referenceToResource
    this.regions = props.regions

    const value = props.resourceToReference(props.resource)

    for (const region of props.regions) {
      new CrossRegionParameter(this, `Param${region}`, {
        name: this.parameterName,
        region,
        value,
      })
    }
  }

  /**
   * Get the resource by resolving the value from SSM Parameter Store
   * in case we are cross-region.
   */
  public get(scope: cdk.Construct, id: string): T {
    const producerRegion = cdk.Stack.of(this).region
    const consumerRegion = cdk.Stack.of(scope).region

    // Fast-path: Same region.
    if (producerRegion === consumerRegion) {
      return this.resource
    }

    if (!this.regions.includes(consumerRegion)) {
      throw new Error(
        `The region ${consumerRegion} is not registered for the parameter`,
      )
    }

    scope.node.addDependency(this)

    new cdk.CfnParameter(scope, `${id}Nonce`, {
      default: this.nonce,
    })

    const reference = ssm.StringParameter.valueForStringParameter(
      scope,
      this.parameterName,
    )
    return this.referenceToResource(scope, id, reference)
  }
}
