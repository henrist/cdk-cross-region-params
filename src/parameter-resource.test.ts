import { Bucket } from "@aws-cdk/aws-s3"
import { App, Stack } from "@aws-cdk/core"
import "jest-cdk-snapshot"
import { ParameterResource } from "./parameter-resource"

test("ssm-parameter-backed-resource in same region", () => {
  const app = new App()
  const stack1 = new Stack(app, "Stack1", {
    env: {
      region: "eu-west-1",
    },
  })

  const bucket = new Bucket(stack1, "Bucket")
  const s = new ParameterResource(stack1, "BucketParam", {
    nonce: "123",
    parameterName: "/my-param",
    // eslint-disable-next-line @typescript-eslint/unbound-method
    referenceToResource: Bucket.fromBucketName,
    resource: bucket,
    resourceToReference: (resource) => resource.bucketName,
    regions: ["eu-central-1", "us-east-1"],
  })

  const stack2 = new Stack(app, "Stack2", {
    env: {
      region: "eu-west-1",
    },
  })

  s.get(stack2, "Bucket")

  expect(stack1).toMatchCdkSnapshot()
  expect(stack2).toMatchCdkSnapshot()
})

test("ssm-parameter-backed-resource in different region", () => {
  const app = new App()
  const stack1 = new Stack(app, "Stack1", {
    env: {
      region: "eu-west-1",
    },
  })

  const bucket = new Bucket(stack1, "Bucket")
  const s = new ParameterResource(stack1, "BucketParam", {
    nonce: "123",
    parameterName: "/my-param",
    // eslint-disable-next-line @typescript-eslint/unbound-method
    referenceToResource: Bucket.fromBucketName,
    resource: bucket,
    resourceToReference: (resource) => resource.bucketName,
    regions: ["eu-central-1"],
  })

  const stack2 = new Stack(app, "Stack2", {
    env: {
      region: "eu-central-1",
    },
  })

  s.get(stack2, "Bucket")

  expect(stack1).toMatchCdkSnapshot()
  expect(stack2).toMatchCdkSnapshot()
})
