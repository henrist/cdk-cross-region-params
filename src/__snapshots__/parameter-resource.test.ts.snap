// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssm-parameter-backed-resource in different region 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Bucket83908E77": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketParamParameucentral1ResoureB1DD6C9F": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "BucketParamParameucentral1ResoureCustomResourcePolicyEAAE68EC",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/my-param\\",\\"Value\\":\\"",
              Object {
                "Ref": "Bucket83908E77",
              },
              "\\",\\"Type\\":\\"String\\",\\"Overwrite\\":true},\\"region\\":\\"eu-central-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my-param\\"}}",
            ],
          ],
        },
        "Delete": "{\\"service\\":\\"SSM\\",\\"action\\":\\"deleteParameter\\",\\"parameters\\":{\\"Name\\":\\"/my-param\\"},\\"region\\":\\"eu-central-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my-param\\"}}",
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/my-param\\",\\"Value\\":\\"",
              Object {
                "Ref": "Bucket83908E77",
              },
              "\\",\\"Type\\":\\"String\\",\\"Overwrite\\":true},\\"region\\":\\"eu-central-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my-param\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "BucketParamParameucentral1ResoureCustomResourcePolicyEAAE68EC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:PutParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "ssm:DeleteParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketParamParameucentral1ResoureCustomResourcePolicyEAAE68EC",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`ssm-parameter-backed-resource in different region 2`] = `
Object {
  "Parameters": Object {
    "BucketNonce": Object {
      "Default": "123",
      "Type": "String",
    },
    "SsmParameterValuemyparamC96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/my-param",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
}
`;

exports[`ssm-parameter-backed-resource in same region 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersf3d3a3cc7f26921b237eff24fc5dd7aef8f0465a1f376b8f7918eb3d4b3e8797ArtifactHashAAFCA968": Object {
      "Description": "Artifact hash for asset \\"f3d3a3cc7f26921b237eff24fc5dd7aef8f0465a1f376b8f7918eb3d4b3e8797\\"",
      "Type": "String",
    },
    "AssetParametersf3d3a3cc7f26921b237eff24fc5dd7aef8f0465a1f376b8f7918eb3d4b3e8797S3BucketBEE108A9": Object {
      "Description": "S3 bucket for asset \\"f3d3a3cc7f26921b237eff24fc5dd7aef8f0465a1f376b8f7918eb3d4b3e8797\\"",
      "Type": "String",
    },
    "AssetParametersf3d3a3cc7f26921b237eff24fc5dd7aef8f0465a1f376b8f7918eb3d4b3e8797S3VersionKeyA877E3C9": Object {
      "Description": "S3 key for asset version \\"f3d3a3cc7f26921b237eff24fc5dd7aef8f0465a1f376b8f7918eb3d4b3e8797\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersf3d3a3cc7f26921b237eff24fc5dd7aef8f0465a1f376b8f7918eb3d4b3e8797S3BucketBEE108A9",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf3d3a3cc7f26921b237eff24fc5dd7aef8f0465a1f376b8f7918eb3d4b3e8797S3VersionKeyA877E3C9",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf3d3a3cc7f26921b237eff24fc5dd7aef8f0465a1f376b8f7918eb3d4b3e8797S3VersionKeyA877E3C9",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Bucket83908E77": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketParamParameucentral1ResoureB1DD6C9F": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "BucketParamParameucentral1ResoureCustomResourcePolicyEAAE68EC",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/my-param\\",\\"Value\\":\\"",
              Object {
                "Ref": "Bucket83908E77",
              },
              "\\",\\"Type\\":\\"String\\",\\"Overwrite\\":true},\\"region\\":\\"eu-central-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my-param\\"}}",
            ],
          ],
        },
        "Delete": "{\\"service\\":\\"SSM\\",\\"action\\":\\"deleteParameter\\",\\"parameters\\":{\\"Name\\":\\"/my-param\\"},\\"region\\":\\"eu-central-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my-param\\"}}",
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/my-param\\",\\"Value\\":\\"",
              Object {
                "Ref": "Bucket83908E77",
              },
              "\\",\\"Type\\":\\"String\\",\\"Overwrite\\":true},\\"region\\":\\"eu-central-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my-param\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "BucketParamParameucentral1ResoureCustomResourcePolicyEAAE68EC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:PutParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "ssm:DeleteParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketParamParameucentral1ResoureCustomResourcePolicyEAAE68EC",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BucketParamParamuseast1ResoureCustomResourcePolicyE00D6F85": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:PutParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "ssm:DeleteParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketParamParamuseast1ResoureCustomResourcePolicyE00D6F85",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BucketParamParamuseast1ResoureEC3872F3": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "BucketParamParamuseast1ResoureCustomResourcePolicyE00D6F85",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/my-param\\",\\"Value\\":\\"",
              Object {
                "Ref": "Bucket83908E77",
              },
              "\\",\\"Type\\":\\"String\\",\\"Overwrite\\":true},\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my-param\\"}}",
            ],
          ],
        },
        "Delete": "{\\"service\\":\\"SSM\\",\\"action\\":\\"deleteParameter\\",\\"parameters\\":{\\"Name\\":\\"/my-param\\"},\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my-param\\"}}",
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/my-param\\",\\"Value\\":\\"",
              Object {
                "Ref": "Bucket83908E77",
              },
              "\\",\\"Type\\":\\"String\\",\\"Overwrite\\":true},\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my-param\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`ssm-parameter-backed-resource in same region 2`] = `Object {}`;
